aws/resource_aws_ec2_client_vpn_endpoint_test.go

```diff
@@ -0,0 +1,382 @@
+	return fmt.Sprintf(`
+resource "tls_private_key" "example" {
+  algorithm = "RSA"
+}
+
+resource "tls_self_signed_cert" "example" {
+  key_algorithm   = "RSA"
+  private_key_pem = "${tls_private_key.example.private_key_pem}"
+
+  subject {
+    common_name  = "example.com"
+    organization = "ACME Examples, Inc"
+  }
+
+  validity_period_hours = 12
+
+  allowed_uses = [
+    "key_encipherment",
+    "digital_signature",
+    "server_auth",
+  ]
+}
+
+resource "aws_acm_certificate" "cert" {
+  private_key      = "${tls_private_key.example.private_key_pem}"
+  certificate_body = "${tls_self_signed_cert.example.cert_pem}"
+}
+
+resource "aws_cloudwatch_log_group" "lg" {
+  name = "terraform-testacc-clientvpn-loggroup-%s"
+}
+
+resource "aws_cloudwatch_log_stream" "ls" {
+  name           = "${aws_cloudwatch_log_group.lg.name}-stream"
+  log_group_name = "${aws_cloudwatch_log_group.lg.name}"
+}
+
+resource "aws_ec2_client_vpn_endpoint" "test" {
+  description = "terraform-testacc-clientvpn-%s"
+  server_certificate_arn = "${aws_acm_certificate.cert.arn}"
+  client_cidr_block = "10.0.0.0/16"
+
+  authentication_options {
+    type = "certificate-authentication"
+    root_certificate_chain_arn = "${aws_acm_certificate.cert.arn}"
+  }
+
+  connection_log_options {
+    enabled = true
+    cloudwatch_log_group = "${aws_cloudwatch_log_group.lg.name}"
+    cloudwatch_log_stream = "${aws_cloudwatch_log_stream.ls.name}"
+  }
+}
+`, rName, rName)
+}
+
+func testAccEc2ClientVpnEndpointConfigWithDNSServers(rName string) string {
+	return fmt.Sprintf(`
+resource "tls_private_key" "example" {
+  algorithm = "RSA"
+}
+
+resource "tls_self_signed_cert" "example" {
+  key_algorithm   = "RSA"
+  private_key_pem = "${tls_private_key.example.private_key_pem}"
+
+  subject {
+    common_name  = "example.com"
+    organization = "ACME Examples, Inc"
+  }
+
+  validity_period_hours = 12
+
+  allowed_uses = [
+    "key_encipherment",
+    "digital_signature",
+    "server_auth",
+  ]
+}
+
+resource "aws_acm_certificate" "cert" {
+  private_key      = "${tls_private_key.example.private_key_pem}"
+  certificate_body = "${tls_self_signed_cert.example.cert_pem}"
+}
+
+resource "aws_ec2_client_vpn_endpoint" "test" {
+  description = "terraform-testacc-clientvpn-%s"
+  server_certificate_arn = "${aws_acm_certificate.cert.arn}"
+  client_cidr_block = "10.0.0.0/16"
+
+  dns_servers = ["8.8.8.8", "8.8.4.4"]
+
+  authentication_options {
+    type = "certificate-authentication"
+    root_certificate_chain_arn = "${aws_acm_certificate.cert.arn}"
+  }
+
+  connection_log_options {
+    enabled = false
+  }
+}
+`, rName)
+}
+
+func testAccEc2ClientVpnEndpointConfigWithMicrosoftAD(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_vpc" "test" {
+  cidr_block  = "10.0.0.0/16"
+}
+
+resource "aws_subnet" "test1" {
+  vpc_id     				= "${aws_vpc.test.id}"
+  cidr_block 				= "10.0.1.0/24"
+  availability_zone = "us-east-1a"
```

Comment:
While configurable, by default the acceptance testing runs in us-west-2:

https://github.com/terraform-providers/terraform-provider-aws/blob/03eb00938807d553711008e93dacfa0c6b0492fa/aws/provider_test.go#L183-L189

Setting these to us-east-1X availability zones generates an error in the acceptance testing:

```
--- FAIL: TestAccAwsEc2ClientVpnEndpoint_msAD (4.15s)
    testing.go:538: Step 0 error: Error applying: 2 errors occurred:
        	* aws_subnet.test2: 1 error occurred:
        	* aws_subnet.test2: Error creating subnet: InvalidParameterValue: Value (us-east-1d) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: us-west-2a, us-west-2c, us-west-2b.
        	status code: 400, request id: 62334444-44e2-4655-8adc-00fde6a2b7a3
        
        
        	* aws_subnet.test1: 1 error occurred:
        	* aws_subnet.test1: Error creating subnet: InvalidParameterValue: Value (us-east-1a) for parameter availabilityZone is invalid. Subnets can currently only be created in the following availability zones: us-west-2a, us-west-2c, us-west-2b.
        	status code: 400, request id: 8218ec9d-a281-49cf-998a-1deef24ab6b9
```

Rather than hardcoding values or hardcoding a provider configuration to the `us-east-1` region unnecessarily, we generally use the `aws_availability_zones` data source to find available availability zones, e.g.

```hcl
data "aws_availability_zones" "available" {}
# ...
  availability_zone = "${data.aws_availablility_zones.available.names[0]}"
# ...
  availability_zone = "${data.aws_availablility_zones.available.names[1]}"
```

Reply:
Fixed.